pkg/nodeagent/controller/lease/add.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2023 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lease
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// ControllerName is the name of the controller.
+const ControllerName = "lease"
+
+// AddToManager adds the lease controller with the default Options to the manager.
+func (r *Reconciler) AddToManager(mgr manager.Manager) error {
+	if r.Client == nil {
+		r.Client = mgr.GetClient()
+	}
+	if r.RenewIntervalSeconds == 0 {
+		r.RenewIntervalSeconds = 60
+	}
+	if r.Clock == nil {
+		r.Clock = clock.RealClock{}
+	}
+	if r.Namespace == "" {
+		r.Namespace = metav1.NamespaceSystem
+	}
+	r.leaseName = "gardener-node-agent-" + r.HostName
+
+	node := &metav1.PartialObjectMetadata{}
+	node.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("Node"))
+
+	return builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(node, builder.WithPredicates(r.NodePredicate())).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 1}).
+		Complete(r)
+}
+
+// NodePredicate returns 'true' when the node controller by node-agent was created.
+func (r *Reconciler) NodePredicate() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return e.Object.GetName() == r.HostName
+		},
+		UpdateFunc:  func(_ event.UpdateEvent) bool { return false },
+		DeleteFunc:  func(_ event.DeleteEvent) bool { return false },
+		GenericFunc: func(_ event.GenericEvent) bool { return false },
+	}
+}
```

Comment:
```suggestion
```

Reply:
This predicate should prevent the reconciler to update the lease on node updates. Create event should start the reconciler, then it should be reconciled after `RenewIntervalSeconds` only.
Without the predicate, frequent updates of the node lead to frequent updates of the lease.

Reply:
This can be easily achieved like this: https://github.com/gardener/gardener/blob/328f202a312de139bb9ae80378c3641c254850c3/pkg/controllermanager/controller/event/add.go#L43

Reply:
In my opinion it is to fragile to rely on the cache only. `node-agent` controllers (lease, node, healthcheck) are designed to reconcile one node only and will start to do weird stuff, in case they would reconcile more than one node.  
`lease` controller could create its leases with OwnerReferences to the wrong node, so that they are garbage collected at the wrong time.
`node` controller could react on annotations of the wrong node, so a restart annotation could cause a restart of a systemd unit at a random node.
`healthcheck` controller could even add its own internal IP to a different node status.

Without predicates for the node names (better labels) you don't see that, when you look at the controllers. 
You and I know that now, but it could be forgotten e.g. when the `controller-runtime` cache implementation is changed again in a year or two. No test would discover that, because the managers of integration tests have a dedicated cache configuration.
There won't be a measurable increase of compute resource. Of course, it adds a couple of code lines, but those will also document, that those are single node controllers.

Reply:
I cannot follow this argument. If we have the tools at hand (like configuring the cache), not using them because "they could change in a year or two" doesn't make much sense to me (unless they are clearly stated as "fragile"/subject to change by the authors - which isn't the case for the manager caches).
If you want, feel free to write a small integration test that ensures that the cache works as expected, but TBH, I don't think this is necessary/over-engineered.
What makes you think the controller-runtime would change their cache implementation in such a breaking way that we cannot adapt to it?

Reply:
Ah, and one point: We don't know the node name upfront for new VMs. We only know the hostname, but when kubelet creates the `Node`, its name could differ. Hence, when really wanted to introduce such predicates, we would need to wait for the `OperatingSystemConfig`s first successful reconciliation (which installs `kubelet` for the first time) and only then register the other controllers (with predicate of correct node name) to the manager. Not a big deal, but definitely complexity and a special code path. I would really like to avoid all this.

Reply:
We could use the [nodeName](https://github.com/gardener/gardener/blob/328f202a312de139bb9ae80378c3641c254850c3/cmd/gardener-node-agent/app/app.go#L150) `node-agent` determines during startup for the predicates too.

In my opinion the cache configuration is meant to configure the cache in an efficient way (e.g. to consume as little memory, CPU as possible). We should definitely do that in `node-agent`.
However, the controller itself should be responsible to ensure that its business logic works properly. This also means that it restricts the objects it is reconciling to those it is responsible for. There will be more code, but it will be more resilient towards undesired side-effects.

I don't want to test `controller-runtime`. This is just an example for something which might go wrong in the future. It could also happen when refactoring `node-agent`. I'm pretty sure that we could adapt our implementation in those cases. The question is "Do we always remember that we have to adapt our implementation?". I cannot name a specific event when (or if) this happens, it is a risk though.

Reply:
> We could use the [nodeName](https://github.com/gardener/gardener/blob/328f202a312de139bb9ae80378c3641c254850c3/cmd/gardener-node-agent/app/app.go#L150) `node-agent` determines during startup for the predicates too.

We can't - this has changed in the branch where gardenlet integrates GNA into the shoot flow, see [here](https://github.com/rfranzke/gardener/commit/9c6dc8247c2b988b90cb32ee3fe52bde4d846db7#diff-55fb15b9da09f83f4983b2003c189b5e2ad4d71afb010bbf61f97dfe1a0b8fcf) the changes in GNA's `app.go` file. When it starts, the node name is not known yet.

> In my opinion the cache configuration is meant to configure the cache in an efficient way (e.g. to consume as little memory, CPU as possible). We should definitely do that in `node-agent`. However, the controller itself should be responsible to ensure that its business logic works properly. This also means that it restricts the objects it is reconciling to those it is responsible for. There will be more code, but it will be more resilient towards undesired side-effects.

The cache options can be used to restrict which objects the component can see. If there is only one object (or a set of objects) relevant to the component which can be perfectly filtered already here, why should we extend each and every controller to double-check again for the "right/expected" set? I really cannot comprehend this argumentation.

We already follow this approach in other components like `gardenlet`, e.g. when it filters for resources only relevant for the seed it is responsible for. [Here](https://github.com/gardener/gardener/blob/524084c5fb31a9b4e4642273f10e96c3e9a2badf/cmd/gardenlet/app/app.go#L262-L264), for example, `ControllerInstallation`s are already filtered based on the `.spec.seedRef.name`, so that we don't need any [additional predicates](https://github.com/gardener/gardener/blob/524084c5fb31a9b4e4642273f10e96c3e9a2badf/pkg/gardenlet/controller/controllerinstallation/controllerinstallation/add.go#L61-L64) in the respective controller(s).

> I don't want to test `controller-runtime`. This is just an example for something which might go wrong in the future. It could also happen when refactoring `node-agent`. I'm pretty sure that we could adapt our implementation in those cases. The question is "Do we always remember that we have to adapt our implementation?". I cannot name a specific event when (or if) this happens, it is a risk though.

Mistakes can and will happen either way. We could also "accidentally" drop the predicate and corresponding tests during a refactoring. With the same probability as newer versions of `controller-runtime` could introduce bugs in the cache options, they could also introduce bugs in the predicate handling. In the end, we rely on this third-party component and must trust that it works as expected. If we don't have this trust, we cannot safely make use of its features, and the logical consequence is then to rewrite the critical parts on our own. I guess nobody wants that.

In summary, the main argument for me is: If we can already filter the relevant objects globally via cache options, there is neither need nor benefit for doing it again and again in each and every single controller. In GNA, we already have two (and soon three) controllers reconciling `Node` objects, and I don't see value in duplicating such check.
Risk that people break stuff during code changes exists always. If you want to be extra sure, you can add a unit test for the `app.go` file that asserts that the manager is started with the expected cache options, but for me this is already far beyond the "tangible". Hence, I vote for keeping it simple and straight-forward.

Reply:
> > We could use the [nodeName](https://github.com/gardener/gardener/blob/328f202a312de139bb9ae80378c3641c254850c3/cmd/gardener-node-agent/app/app.go#L150) `node-agent` determines during startup for the predicates too.
> 
> We can't - this has changed in the branch where gardenlet integrates GNA into the shoot flow, see [here](https://github.com/rfranzke/gardener/commit/9c6dc8247c2b988b90cb32ee3fe52bde4d846db7#diff-55fb15b9da09f83f4983b2003c189b5e2ad4d71afb010bbf61f97dfe1a0b8fcf) the changes in GNA's `app.go` file. When it starts, the node name is not known yet.

Here is the PR that adapts the handling regarding the node determination: https://github.com/gardener/gardener/pull/8826

Btw, I saw that we already rely on the cache options for the existing `Node` controller. We don't any predicate regarding the "correct" node there: https://github.com/gardener/gardener/blob/eba989bc166232707bb1309c7df61cc6fa1ee8ca/pkg/nodeagent/controller/node/add.go#L47-L69

I hope you are okay with continuing with this approach as motivated/justified in the previous comment.